src/vcpkg/base/messages.cpp

```diff
@@ -95,17 +96,160 @@ namespace vcpkg
 
     LocalizedString::LocalizedString(StringView data) : m_data(data.data(), data.size()) { }
     LocalizedString::LocalizedString(std::string&& data) noexcept : m_data(std::move(data)) { }
+
 }
 
 namespace vcpkg::msg
 {
-    REGISTER_MESSAGE(SeeURL);
-    REGISTER_MESSAGE(NoteMessage);
-    REGISTER_MESSAGE(WarningMessage);
-    REGISTER_MESSAGE(ErrorMessage);
-    REGISTER_MESSAGE(InternalErrorMessage);
-    REGISTER_MESSAGE(InternalErrorMessageContact);
+    namespace
+    {
+        template<class T>
+        struct ArgExample;
+    }
+}
+#define DECLARE_MSG_ARG(NAME, EXAMPLE)                                                                                 \
+    StringLiteral vcpkg::msg::NAME##_t::name = #NAME;                                                                  \
+    template<>                                                                                                         \
+    struct vcpkg::msg::ArgExample<::vcpkg::msg::NAME##_t>                                                              \
+    {                                                                                                                  \
+        static constexpr StringLiteral example = sizeof(EXAMPLE) > 1 ? StringLiteral("{" #NAME "} is " EXAMPLE ".")    \
+                                                                     : StringLiteral("");                              \
+    };
+#include <vcpkg/base/message-args.inc.h>
+#undef DECLARE_MSG_ARG
+
+namespace vcpkg
+{
+    namespace
+    {
+        static constexpr const size_t max_number_of_args = 5;
+
+        struct MessageData
+        {
+            StringLiteral name;
+            std::array<const StringLiteral*, max_number_of_args> arg_examples;
+            StringLiteral comment;
+            StringLiteral builtin_message;
+        };
+
+        template<class... Args>
+        constexpr std::array<const StringLiteral*, max_number_of_args> make_arg_examples_array(Args...)
+        {
+            return std::array<const StringLiteral*, max_number_of_args>{&msg::ArgExample<Args>::example...};
+        }
+
+        constexpr MessageData message_data[] = {
+#define DECLARE_MESSAGE(NAME, ARGS, COMMENT, ...) {#NAME, make_arg_examples_array ARGS, COMMENT, __VA_ARGS__},
+#include <vcpkg/base/message-data.inc.h>
+#undef DECLARE_MESSAGE
+        };
+    }
+
+    namespace msg::detail
+    {
+        static constexpr const ::size_t number_of_messages = sizeof(message_data) / sizeof(message_data[0]);
```

Comment:
```suggestion
        static constexpr ::size_t number_of_messages = sizeof(message_data) / sizeof(message_data[0]);
```

Alternately, should this be `std::size`?